### DISCUSSION TITLE
Custom OAuth provider

### DISCUSSION THREAD

**Discussion by Aragur** [pocketbase/pocketbase#460] [Q&A] [answered]
Hey, is there some way to integrate a custom OAuth provider?
Currently I'm using [authenthik](https://goauthentik.io/) as a self hosted Auth-Provider for SSO across all applications.

**Comment by ganigeorgiev** [comment, marked as answer]
Hm, I'm not quite sure that I understand the request here.

Authentik doesn't look like a regular OAuth2 provider, but an auth platform  that has its own OAuth2/SAML/LDAP/etc. providers integration.
If you want to "forward" the authenticated user from _authentik_ to PocketBase you cannot do that directly because you'll have to create somehow the user in PocketBase beforehand and this is too user-land specific.
When PocketBase is used as framework you can register a custom middleware and use for example a trusted header value to find and authenticate the user in PocketBase. For example, you could check https://github.com/pocketbase/pocketbase/issues/136#issuecomment-1186191919.

---

PocketBase currently support only external OAuth2 providers but they cannot be added dynamically because when adding a new OAuth2 provider usually this also require some additional changes in the Admin UI which is not extandable at the moment. There are plans in the future to move the frontend provider options to be entirely configurable from server-side so that you'll be able to add any custom OAuth2 provider when PocketBase is used as framework.

**Reply by ganigeorgiev** [reply]
Actually, it may be possible to use authentik as a regular OAuth2 provider that will just return back to PocketBase the authenticated user data after its own auth flow, but their [api documentation](https://goauthentik.io/api/) is a little confusing and I'll have to research it more thoroughly whether it would be possible to use it this way.

_In any case new OAuth2 providers will be added after the users and profiles refactoring in #376 and the "Apple sign-in" will be prioritized because it may be a blocker for users working on ios applications due to the 3rd party auth requirements of the app store listed in https://github.com/pocketbase/pocketbase/issues/202#issuecomment-1242652375._

**Reply by Aragur** [reply]
@ganigeorgiev Hey, so now that version 0.8 is released I was thinking about a possibility to implement a custom OAuth provider field in the settings, where you can input the Base-Domain of your Auth provider.
I also implemented the OAuth provider on the [Appwrite Project](https://github.com/appwrite/appwrite/pull/3180)

**Reply by ganigeorgiev** [reply]
@Aragur I'm not sure that I understand your proposal. Each OAuth2 provider needs:

- defined scopes
- auth url (the url to the provider's consent page that asks for permissions based on the scopes)
- token url (the url to exchange the auth code for a token)
- user/profile url (the url to the provider api that uses the exchanged token to get user/profile info like account id, email, avatar, etc.)
- spacial function that will take care to map the the provider user's fields to the PocketBase's `auth.AuthUser` struct

The first 4 could be added in the Admin UI as an option control, but the last depends on the provider's api and cannot be automated (some providers, eg. GitHub, requires several request to get all required fields).

**Reply by Aragur** [reply]
@ganigeorgiev Ah okay I get the issue.
So it would be better to implement Authentik directly?

**Reply by Aragur** [reply]
A second thought as in my use case I need a single sign on for other applications.
Do you think implementing a OAuth provider for Pocketbase would be a lot of work?
So that I can add Pocketbase as a SignIn Option to other applicaitons via OAuth?

**Reply by ganigeorgiev** [reply]
> So it would be better to implement Authentik directly?

Yes. If you want to contribute, you can check for a simple example the recent Spotify PR - https://github.com/pocketbase/pocketbase/pull/887/files (you can ignore the ui/dist files).

---

> Do you think implementing a OAuth provider for Pocketbase would be a lot of work? So that I can add Pocketbase as a SignIn Option to other applicaitons via OAuth?

I'm not sure that I understand what you are asking but PocketBase doesn't have OAuth2 api.

**Reply by Aragur** [reply]
> I'm not sure that I understand what you are asking but PocketBase doesn't have OAuth2 api.

Yes, I was talking about the possibility to add this feature :)

**Reply by ganigeorgiev** [reply]
At least for v1.0.0 there are no plans to implement an OAuh2 api because it will complicate too much the existing authentication flow (for example, in PocketBase there is no concept for "scopes" since the permissions and access controls depends entirely on the collection API rules; the code, token and refresh token exchange flow is also very different than what we currently do).
